src/fs/fs.ts

```diff
@@ -358,18 +388,33 @@ export class FileSystem extends EventEmitter implements FlashDataSource {
 
   async statistics(): Promise<Statistics> {
     const fs = await this.initialize();
-    const currentMainFile = fs.read(MAIN_FILE);
+    const currentMainFile = fs.readBytes(MAIN_FILE);
     return {
       files: fs.ls().length,
       storageUsed: fs.getStorageUsed(),
       lines:
         this.cachedInitialProject &&
-        this.cachedInitialProject.main === currentMainFile
+        this.cachedInitialProject.files[MAIN_FILE] ===
+          fromByteArray(currentMainFile)
           ? undefined
-          : currentMainFile.split(/\r\n|\r|\n/).length,
+          : this.lineNumFromUint8Array(currentMainFile),
     };
   }
 
+  lineNumFromUint8Array(arr: Uint8Array): number {
+    let lineCount = 1;
+    let prevByte: number | undefined;
+    const LF = 10;
+    const CR = 13;
+    arr.forEach((byte) => {
+      if ((byte === LF && prevByte !== CR) || byte === CR) {
+        lineCount++;
+      }
+      prevByte = byte;
+    });
+    return lineCount;
+  }
```

Comment:
I think I'd extract this to its own function and write a simple unit test for it, as it could easily be wrong and we wouldn't really notice. I think it's right though 😄

Reply:
Extracted and test added here - [`23fcb8d`](https://github.com/microbit-foundation/python-editor-next/pull/626/commits/23fcb8d67ddb9653d8d439a87b42c76a59853112)

Reply:
Looks good, thanks.